### DISCUSSION TITLE
Failing to receive using interrupt method with ESP32 and SX1268 (NiceRF module)

### DISCUSSION THREAD

**Discussion by strud** [jgromes/RadioLib#593] [Q&A]
I have a project with an airborne telemetry board and ground station, both using ESP32 modules and Nice RF modules with SX1268 + PA.

I am hapily able to connect and configure both radios and I beleive their settings should make them able to communicate.

The airborne device has been set to tranmit a LORA packet every 5s and I can confirm on my SDR that it seems to look like it is on freq and 'looks' reasonable.

The ground station end is setup to receive as per the interrupt example but I never get a reception event.

I note that these Nice RF modules only expose DIO1 and DIO3. Maybe DIO2 is used internally for the antenna/RF switch?

I have now tried using the function setDio2AsRfSwitch() but this has not resolved the issue.

I have tried changing the constructor to use DIO3 and the 4th argument instead of the BUSY pin - this results in a error 707

Open to suggestions?

**Comment by jgromes** [comment]
> I have tried changing the constructor to use DIO3 and the 4th argument instead of the BUSY pin - this results in a error 707

The 4th argument of the constructor has to be the BUSY pin. DIO3 is typically used as a reference voltage source for TCXO. DIO2 is also commonly used to control RF switch - so commonly in fact, that RadioLib enables it automatically, hence no need to call setDio2AsRfSwitch().

Unfrortunately it's not possible to deduce the cause unless you post the relevant information - codes for transmitter and receiver, links to the hardware you mentioned, wiring diagram etc.

**Reply by strud** [reply]
Hi Jan,

Instantiation for both ends:

Rx:   `SX1268` radioModem = new Module (VCS2, DIO1, NRESET,BUSY, SPI, SPISettings(2000000, MSBFIRST, SPI_MODE3)); `
Tx    `SX1262 radioModem = new Module(VCS,DIO1, NRESET,BUSY, SPI, RADIOLIB_DEFAULT_SPI_SETTINGS);`

Tx parameter setting:
`radioState = radioModem.begin(430.0F,250.0, 10, 5, 0x34, 22,20,0,false);
  digitalWrite(ARM_STATUS, 0);
  if (radioState == RADIOLIB_ERR_NONE) {
      Serial.println(F("success!"));
      radioConfigure(&localSystem.radioModemSettings, &radioModem);
      radioModem.setCurrentLimit(500);
    } else {
      Serial.print(F("failed, code "));
      Serial.println(radioState);
      //while (true);
   }`

Rx Parameter setting:
`Serial.print(F("[SX1268] Initializing"));
  u8g2log.print("radio modem-");
  deassertLcdCS //  make sure screen CS low (not asserted) before talking to radio
  delay(5);
  
  radioState = radioModem.begin(430.0F,250.0, 10, 5, 0x34, 22,20,0,false);
  deassertRadioCS //  make sure radio CS is high before talking to screen
  if (radioState == RADIOLIB_ERR_NONE) {
      Serial.println(F("success!"));
      u8g2log.println("ok");
      peripheralStatus.radioModemDetected=1;
      peripheralStatus.radioModemOperational=1;
      deassertLcdCS //  make sure screen CS low (not asserted) before talking to radio
      delay(5);
      radioConfigure(&localSystem.radioModemSettings, &radioModem);
      deassertRadioCS // make sure radio CS is high before talking to screen
      } 
  else {
        Serial.print(F("failed, code "));
        Serial.println(radioState);
        u8g2log.println("BAD");
        peripheralStatus.radioModemDetected=0;
        peripheralStatus.radioModemOperational=0;
        //while (true);
     }`

The common radio parameter setting routine:

`int radioConfigure(radioModemType *modemParams, SX1268 *radio){
  // you can also change the settings at runtime
  // and check if the configuration was changed successfully
  int error=0;
  // set carrier frequency
  if (radio->setFrequency(modemParams->channelFrequency) == RADIOLIB_ERR_INVALID_FREQUENCY) {
    Serial.println(F("Selected frequency is invalid for this module!"));
    error++;
  }
  // set bandwidth
  if (radio->setBandwidth(modemParams->bandWidth) == RADIOLIB_ERR_INVALID_BANDWIDTH) {
    Serial.println(F("Selected bandwidth is invalid for this module!"));
    error++;
  }
  // set spreading factor
  if (radio->setSpreadingFactor(modemParams->spreadingFactor) == RADIOLIB_ERR_INVALID_SPREADING_FACTOR) {
    Serial.println(F("Selected spreading factor is invalid for this module!"));
    error++;
  }
  // set coding rate 
    if (radio->setCodingRate(modemParams->codingRate) == RADIOLIB_ERR_INVALID_CODING_RATE) {
    Serial.println(F("Selected coding rate is invalid for this module!"));
    error++;
  }
  // set LoRa sync word
  if (radio->setSyncWord(modemParams->synchWord) != RADIOLIB_ERR_NONE) {
    Serial.println(F("Unable to set sync word!"));
    error++;
  }
  // set output power (accepted range is -17 - 22 dBm)
  if (radio->setOutputPower(modemParams->txPower) == RADIOLIB_ERR_INVALID_OUTPUT_POWER) {
    Serial.println(F("Selected output power is invalid for this module!"));
    error++;
  }
  // set over current protection limit to 80 mA (accepted range is 45 - 240 mA)
  // NOTE: set value to 0 to disable overcurrent protection
  if (radio->setCurrentLimit(120) == RADIOLIB_ERR_INVALID_CURRENT_LIMIT) {
    Serial.println(F("Selected current limit is invalid for this module!"));
    error++;
  }
  // set LoRa preamble length (accepted range is 0 - 65535)
  if (radio->setPreambleLength(15) == RADIOLIB_ERR_INVALID_PREAMBLE_LENGTH) {
    Serial.println(F("Selected preamble length is invalid for this module!"));
    error++;
  }
  if (radio->setDio2AsRfSwitch() != RADIOLIB_ERR_NONE) {
    Serial.println(F("Failed to set DIO2 as RF switch!"));
    error++;
  }
  return error;
}`



The telemetry state machine in the Rx end:

`    // radio comms/telemetry state machine
      switch (stateRadio){

      case st_rad_initial:
          // initial connection to radio is done in setup()
          // check that the result was success
          if ((peripheralStatus.radioModemDetected==1)&&(peripheralStatus.radioModemOperational==1)){
             stateRadio = st_rad_configure; 
              Serial.println("->st_rad_configure");
          }
          else {
            stateRadio = st_rad_error; 
            Serial.println("->st_rad_error");
          }


      break;

      case st_rad_initialise: // seems like there was a problem with original initialisation

      break;

      case st_rad_configure:
          if (spiMutex == false){
              spiMutex=true;
              deassertLcdCS
              radioModem.setDio1Action(isr_setRXFlag);  // set the function that will be called
                                                                // when new packet is received
              error = radioModem.startReceive();                // put radio into listen mode
              if (error == RADIOLIB_ERR_NONE) {
                stateRadio = st_rad_standby; 
                Serial.println("->st_rad_standby");
              }
              else {
                stateRadio = st_rad_error;
                Serial.println("->st_rad_error");
              }
              spiMutex=false;
              deassertRadioCS;
          }

      break;

      case st_rad_standby:
          // check for a recieved packet
          if (radioRXFlag){
            // something has arrived, time to read and parse it
            enableRxInterrupt = false;
            // reset flag
            radioRXFlag = false;
            stateRadio = st_rad_read_pckt;
            Serial.println("->st_rad_read_pckt");
          }
      break;`

ISR (basically per example)

`//---------------------------------------------------------------------------
// function to service radio RX interrupt
//---------------------------------------------------------------------------
#if defined(ESP8266) || defined(ESP32)
  ICACHE_RAM_ATTR
#endif
void isr_setRXFlag(void){
  // check if the interrupt is enabled
  if(!enableRxInterrupt) {
    return;
  }
  // we got a packet, set the flag
  radioRXFlag = true;
}`

Interrupts turned on end of Setup:
`// Turn on interrupts
  interrupts();
  
}`


So currently, the Tx end seems to be transmitting on 435MHz as set in the localSystem structure (same both ends).

The Rx end reports no errors and the state machine gets to the point of waiting for the interrupt flag but this never gets set.

**Reply by strud** [reply]
Ok forget all of that.

I had not forward declared the ISR. Now that I have done that it is working......... oh dear

**Reply by jgromes** [reply]
Well I'm glad you got it working ;)
